### DISCUSSION TITLE
CUDA/OpenCL interop: Rendering GPU-generated images

### DISCUSSION THREAD

**Discussion by JDr111** [unoplatform/uno#15293] [Q&A] [answered]
Hello,
I need to develop a cross-platform desktop app (Windows, Linux and MacOS only, no mobile), and I'm currently evaluating whether Uno is a suitable GUI stack for my needs.

Essentially, I'm constructing 2D images on the GPU using either CUDA or OpenCL, and I need to display these images in the user interface. Interop between CUDA/OpenCL and OpenGL/Vulkan is viable. This custom "renderer" should provide the ability to zoom and scroll around the image, and apply post-processing effects (such as blur) to the rendered image, without having to resort to software rendering and shuffling the image data between CPU and GPU memory (somehow rendering a textured quad using OpenGL or Vulkan is one way to do it, but any other means is acceptable).

To summarize the question: How would one go about rendering 2D images that already reside in GPU memory in a cross-platform manner?

Thanks!

**Comment by jeromelaban** [comment]
Thanks for the question! It depends on the platform, but at this time, rendering images this way can only be done in a cross-platform manner through rendering in a `byte[]` capture of the opengl (or others) surfaces, then placing it in a `WriteableBitmap`. We're evaluating  https://github.com/unoplatform/uno/issues/9405, but we have no concrete plans yet.

Which platform are you targeting more specifically?

**Reply by JDr111** [reply]
I'm targeting Windows (most important), Linux (important), and Mac OS X (would be great). I do realize supporting Mac OS X will be tricky simply due to the fact Apple no longer supports OpenGL or Vulkan on its newest software/hardware, only its own Metal API. So I do count with the possibility of dropping Mac OS X support, at least for the time being.

**Reply by jeromelaban** [reply]
Thanks for the details. 

On Windows, Uno is not present for the WinAppSDK head, but there's work being done for WinAppSDK by Microsoft to get Angle running (https://github.com/mono/SkiaSharp/issues/1893, https://github.com/microsoft/vcpkg/pull/36181 and https://chromium-review.googlesource.com/c/angle/angle/+/5232157) , which may help in some capacity for what you're trying to reach. 

For Linux or Uno+Skia in general, you can keep track of https://github.com/unoplatform/uno/issues/9405.

**Reply by JDr111** [reply]
Thanks for the information!

I was wondering about a potential alternate solution. This could be a solvable problem if it's possible to embed an external window inside a Uno Platform host. This way, the renderer can be a completely separate DLL project. Does Uno Platform provide such APIs?

**Reply by jeromelaban** [reply, marked as answer]
I may be possible to do this with WinAppSDK, but we're not sure.

Using Uno (wpf/gtk), it is possible to place an additional window as part of the running app. The [mediaplayer support](https://github.com/unoplatform/uno/blob/55eab317c70fe07c8d73191fe2324b1afaef6acc/src/AddIns/Uno.UI.MediaPlayer.Skia.Gtk/GtkMediaPlayer.cs#L24) uses this feature to integrate the video output in the app. It's a somehow manual process, though, as the window has to be resize when the corresponding hosting control is resized.

**Reply by JDr111** [reply]
Thanks for the info, that's a very good pointer. Manual process is not an issue, so long as it's possible to achieve all set goals. I'll investigate the media player support, see if it's adjustable to my needs. Marking the answer, thanks again!
